modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/ClientFailoverSocket.cs

```diff
@@ -269,6 +278,12 @@ private ClientSocket GetAffinitySocket<TKey>(int cacheId, TKey key)
                 return null;
             }
 
+            // Transactional operation should be executed on node started the transaction.
+            if (_transactions.CurrentTx != null)
```

Comment:
To be precise: transactional operation must be executed using the same connection (same `ClientSocket` instance) that started the transaction.

And the check here does not guarantee this at all - reconnect could have happened, for example.
A proper approach is to store get the socket from `ClientContextBase.Socket` property when executing `ClientOp.TxStart`, and use that socket directly for transactional operations.